Allow multiple link shares

PVince81 opened the pull request:
## Description
Removes the artifical restriction in the code that limited link shares to a single entry.

Please note that the limit was put in place because the UI wasn't ready for multiple link shares.
Before merging this, make sure the UI is adjusted first.

## Related Issue

## Motivation and Context
We want multiple link shares!

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

@fheidecke here you go, please use this branch when testing the frontend changes.
Multiple link shares can simply be created by using POST when creating the share and PUT when changing it. Each will get a different id. It's similar to regular user shares.

felixheidecke commented:
Does this support all features shown in this clickdummy?
https://felixheidecke.github.io/owc_clickdummys/pages/sharing.html

PVince81 commented:
> Does this support all features shown in this clickdummy?

very likely

PVince81 commented:
Tasks:

- [ ] create SharesCollection class for shares and use it only for link shares for now
- [ ] create ShareModel class for shares but only use for link shares for now
    - [ ] fetch must get it
    - [ ] save must create or update it
- [ ] make ShareItemModel's `getLinkShare` method return a collection instead of an array

- [ ] create a ShareLinkCollectionView representing the link shares list
    - [ ] based / wired with a SharesCollection (any event rerenders the whole list)
    - [ ] put it in a separate sub-tab "Public links" as per click-dummy
    - [ ] action for opening popup
        - [ ] set a newly created ShareModel
        - [ ] add it to collection if popup was closed with OK
    - [ ] action for deletion (collection.get(shareId).destroy())
    - [ ] action for edition: get the ShareModel from the collection and pass it to the popup

- [ ] repurpose ShareDialogLinkShareView for the popup
    - [ ] remove all saving logic and move it to the model if needed (we only save on OK, not for every field change)
    - [ ] triggers events for OK and Cancel/close
    - [ ] based on a ShareModel
    - [ ] merge expiration subview into this one
    - [ ] put it into a popup dialog (need to find how?)
    - [ ] clicking OK calls `shareModel.save()` and closes

lordelix committed:
Add "Add" and "Remove" button to link share,
remove individual input events in link share,
join link share and expiration in one model

PVince81 commented:
- [ ] TODO: make sure that when resharing is not allowed the UI appear properly accordingly

PVince81 commented:
- [ ] TODO: the share link social view must only appear / be created when clicking the share/anchor icon (only a single view might be needed and append it within the row)

PVince81 commented:
- [ ] TODO: double check expire data inconsistency
- [ ] TODO: test server side errors handling in link dialog (ex: password policy)

Commit:
Split link view to link list and link popup

Non-functional currently...

Commit:
Improve wiring for share link popup

Commit:
Fix clipboard, social share and upload checkbox

PVince81 commented:
@felixheidecke please have a look at my changes, let me know if something is unclear.

We still need the subtab for "Public links".

Regarding the merging of the expiration view I'd recommend against it to save some time. If needed I can simply rewire it for now. Let me know what you think.

PVince81 commented:
- [ ] TODO: email field

Should the email field rather be part of the anchor feature ? (share with social, share with email, copy to clipboard). It feels weird to have an email field in the popup and need to click "Edit" to access it.

PVince81 commented:
- [ ] TODO: adjust unit tests (once the structure is set in stone)

PVince81 commented:
- [ ] only show email field in popup on create
- [ ] test if email field actually works

- [ ] investigate how to pass / add a new "label" field for link naming

PVince81 commented:
- [ ] TEST: reshares and reshare info display

felixheidecke commented:
> It feels weird to have an email field in the pop-up and need to click "Edit" to access it.

As we discussed just now, we need this in the modal to be able to access the password for future usage with E-Mail and Text-Messages (SMS)

Commit:
Split local and public shares in separate subtabs

Commit:
Add "name" field in oc_share

PVince81 commented:
- [ ] REQUIRES: PR to have oc_share be built within a migration: https://github.com/owncloud/core/pull/27394

Then once we have this we can use another migration to add the new column.

PVince81 commented:
I've already added the migration commit + the new name column + passing it through the APIs on this PR.
There's now a name field now in the popup which value is saved/edited properly.

Now if we want the name to be prepopulated to be the token, then it means we need to pre-create the share before opening the popup for editing. If we really want that, then once the share model reaches the popup it isn't a new share any more. So we'll need to pass a flag to the dialog as it cannot rely on `isNew()` any more.

@felixheidecke can you confirm that this is the flow we want ?

PVince81 commented:
- [ ] creation date in tooltip ? (it's the "stime" attribute)

felixheidecke commented:
@PVince81 My suggestion is the following (in code terms): `echo (name) ? name : token`
There is no need to populate the name field if the name is null. Agree?

PVince81 commented:
> My suggestion is the following (in code terms): echo (name) ? name : token
There is no need to populate the name field if the name is null. Agree?

Yes, agree.

felixheidecke commented:
> creation date in tooltip?

Depends on where the tooltip lives. Go ahead and implement it. I can take care of it whenever I bügel glatt the UI.

Commit:
Add mail view and name in list

PVince81 commented:
I'll let you do the share time stuff.

In the last push https://github.com/owncloud/core/pull/27337/commits/a7db59944f94d63ead70a7258747813e1339653b:
- added mail view to the popup (only for new shares)
- email is sent when clicking ok, right after share creation
- removed the link share text field and replaced it with the name (with the logic `name || token`)

@felixheidecke I think you can take over from here. Let me know when the structures are stable enough so I can update most unit tests.

Commit:
Lazy init link view

Commit:
Prevent sidebar tabs to control subtabs

PVince81 commented:
I've just added a few more tweaks:
- link view is now rendered lazily
- link view and subtabs properly disappear when link sharing is globally forbidden
- properly display "resharing is not allowed" on the main view
- prevent sidebar tab switching to affect subtabs (it was a bug in the sidebar code)

PVince81 commented:
- [ ] TODO: server-side unit tests for the newly added "name" column...

Commit:
Fix some sinon warnings about callsFake

Commit:
Adjust and add more unit tests

PVince81 commented:
- [ ] TODO: reimplement password enforcement logic (visual clue about required field + validation on save)

Commit:
Even more tests added for multiple link shares

lordelix committed:
First rough styling of multiple link shares,
add new color versions of some icons

PVince81 commented:
- [ ] BUG: clipboard button doesn't seem to work for hidden fields. Maybe need to hide it offscreen instead

Commit:
Updated tests + added tests for mail

PVince81 commented:
- [ ] little clear button to clear fields, unless the field value is required

Commit:
Fix share link dialog for easier validation

Commit:
Add more error handling for password enforcement

Commit:
Remove obsolete option "Always ask for password" for link shares

Commit:
Only send required attrs for share links

Also remove the password attribute when one is already set and the user
did not want to change it

Commit:
Add back and tweak link expiration view

Commit:
Add explicit sinon lib

Commit:
Fix more share link tests

lordelix committed:
Improve styling

lordelix committed:
Add css-class attribute to dialog message, mute _setSite method

lordelix committed:
New styling of user, group and link shares

lordelix committed:
Revert changes to sharee- list

Commit:
Fix share link icon in link list

PVince81 commented:
We can't merge this yet as there is no proper way for copying the link.
Minimum would be to make the clipboard thing work to avoid losing this in a major way.

- [ ] BUG: no way to copy link

In general I'd also like to have the icon order sorted out already before merging, if possible. Trash is still too close to the share button.

Commit:
Add back linkPassText to unbreak the code and tests

lordelix committed:
Add confirmation to delete link; Add copy-to-clipboard

PVince81 commented on core/js/sharedialoglinklistview.js:
```diff
@@ -107,12 +108,22 @@
+				t('core', 'Delete {link}', { link: linkTitle }),
+				t('core', 'Remove link'),
+				function(cb) {
+					if (cb)
```
please always add the brackets for readability

and also a `// TODO: handle error`

PVince81 commented on apps/files_sharing/css/sharetabview.css:
```diff
@@ -89,17 +89,28 @@
   height: 32px;
   width: 32px;
   background-color: #b4b4b4;
+  background-image: url('/core/img/actions/public-white.svg');
```
please revert back and leave my change, this breaks on every setup where ownCloud is not deployed under "/core"

why not use the "icon-public-white" class like I did ? that one has the correct path.

Also note that we **must** use the icon-* classes because some themes might override these.

And we should probable not use ":before" here but a real element because ":before" cannot be overridden by themes.

lordelix committed:
Revert CSS changes.

lordelix committed:
Add brackets for readability

lordelix committed:
Style social-share icons box

lordelix committed:
Always return the full URL

PVince81 commented:
- [ ] add hook inside the popup to allow apps to extend the dialog's behavior for example with new fields

lordelix committed:
Remove unnecessary comments

Commit:
Tweaks for link shares

Add back anchor icon in list.
Fix generated link.
Added extension point in dialog.

PVince81 commented:
I've added back the anchor icon which disappeared...

Also added extension point in popup dialog.

Fixed ShareModel.getLink() but unit test will fail... I originally changed that method to make the test pass but now it seems it's better to fix the test.

- [ ] Fix ShareModel.getLink() unit test

Commit:
Make OC.webroot properly stubbable

Make other OC.* API calls use OC.getRootPath() instead of directly using
OC.webroot which makes it possible to properly stub OC.getRootPath() for
testing.

Commit:
Fix link share generation tests

lordelix committed:
Password input is type of password

jvillafanez commented on core/js/config.php:
```diff
@@ -70,7 +70,6 @@
 $countOfDataLocation = 0;
 
 $value = $config->getAppValue('core', 'shareapi_enable_link_password_by_default', 'no');
-$enableLinkPasswordByDefault = ($value === 'yes') ? true : false;
```
Is this on purpose?

PVince81 replied:
Ah, yes...

Remember how the link creation used to work before this PR: it was sometimes like a step by step process especially when enforcement is set. Some contributor added this option to make the password field appear first to suggest more strongly to a user to set a password, but the user still had the option to leave it empty. (so it's less strong than "enforce password").

Since we modified the flow by having a dialog, this option becomes obsolete because there is no more step by step process.

jvillafanez replied:
👍

jvillafanez commented on lib/private/Share20/Share.php:
```diff
@@ -417,4 +419,19 @@ public function setMailSend($mailSend) {
+	/**
+	 * @inheritdoc
+	 */
+	public function setName($name) {
```
taking into account the limitation in the DB I'd rather check the length of the string and do something if it's over the limit.

PVince81 replied:
This should rather be done in the DefaultShareProvider because that's the one close to the DB.
From what I see the current DefaultShareProvider's DB call would simply throw if the field is too big, which I think is fine for now... (ideally such exceptions should be encapsulated in a higher level exception type)

Here in this class we're on the abstract level and some providers might allow for bigger fields.

jvillafanez replied:
Yes, it makes more sense to check it in the DefaultShareProvider.
What worries me is that the exception isn't documented. Right now, if I use this function and try to set a big name I expect it works, not throwing an exception. There is nothing telling me "this could fail under these conditions"

PVince81 replied:
ok, will take care here https://github.com/owncloud/core/pull/27448#issuecomment-288359472 once I get Oracle stuff sorted out

jvillafanez commented on lib/public/Share/IShare.php:
```diff
@@ -323,4 +323,20 @@ public function setMailSend($mailSend);
 	public function getMailSend();
+
+	/**
+	 * Set an arbitrary name for this share
```
To be decided how we're going to handle any limitation here. If we state here that the length should be 64 bytes at most, then it's expected both the implementations and anything using them will respect that limitation.
With the current documentation, the implementation have to take care of name over 64 bytes somehow, otherwise it's a bug in the implementation.

jvillafanez commented on lib/public/Share/IShare.php:
```diff
@@ -323,4 +323,20 @@ public function setMailSend($mailSend);
+	public function setName($name);
+
+	/**
+	 * Get arbitrary name for this share
```
This could lead to confusion. An arbitrary name could be any random name, not just the one set previously by `setName`, which is what anyone would expect.

PVince81 replied:
Indeed. I actually meant "user-defined" name. I'll adjust.

jvillafanez commented on tests/lib/Share20/DefaultShareProviderTest.php:
```diff
@@ -1904,6 +1911,8 @@ public function testUpdateLink() {
 		$share->setNode($file2);
 		$share->setPermissions(1);
+		$share->setToken('anothertoken');
+		$share->setName('another_name');
```
it should be possible to set `null` as name, or not to set a name, is this case taken into account?

PVince81 replied:
null as name is possible

felixheidecke commented on core/js/sharedialoglinkshareview.js:
```diff
@@ -68,270 +54,248 @@
-			if(event.keyCode == 13) {
-				this.onPasswordEntered();
+			if (this.configModel.get('enforcePasswordForPublicLink')
+				&& !password.trim()
```
Password isn't trim-ched above (https://github.com/owncloud/core/pull/27337/files#diff-da7df34db49581af77bd59e4f269dc36R97) but here it is. Only spaces and tabs could be valid password?!?!

PVince81 replied:
I'll remove the trim bit.

felixheidecke commented on core/js/sharedialoglinkshareview.js:
```diff
@@ -68,270 +54,248 @@
+				isMailEnabled: showEmailField
 			}));
 
+			this.$('.datepicker').datepicker({dateFormat : 'dd-mm-yy'});
```
See https://github.com/owncloud/core/pull/27337/files#r107128622

felixheidecke commented on core/js/sharedialoglinksocialview.js:
```diff
@@ -30,14 +28,12 @@
-			'	data-url="mailto:?subject=&body={{reference}}"></button>' +
-			'    {{/if}}' +
-			'{{/if}}'
+			'	data-url="mailto:?subject=&body={{reference}}"></button>'
```
`subject` isn't needed if left blank. https://tools.ietf.org/html/rfc2368

PVince81 replied:
copy-pasted from old code, remove ?

PVince81 replied:
will remove it

felixheidecke commented on core/js/sharedialogmailview.js:
```diff
@@ -38,49 +34,61 @@
+		 *
+		 * @param {string} recipientEmail recipient email address
+		 */
+		_sendEmailPrivateLink: function(recipientEmail) {
```
What does 'private' imply here?

PVince81 replied:
No idea... I just copied the old method name.

I think it's still a public link but the email might be what is private...

lordelix committed:
Remove unused local variable

PVince81 commented on apps/files_sharing/lib/API/Share20OCS.php:
```diff
@@ -329,16 +332,6 @@ public function createShare() {
 
-			/*
-			 * For now we only allow 1 link share.
-			 * Return the existing link share if this is a duplicate
```
Hmm good question... I think that what Roeland meant back then was simply "multiple links".
I don't think it's possible to have duplicate links which would mean having duplicate tokens.

jvillafanez replied:
I expect the DB would take care of this, but since we suppose to support several share providers I don't know what component should be responsible of the deduplication or prevent the duplication.

PVince81 replied:
Ok, I'll simply write an additional test here https://github.com/owncloud/core/pull/27448#issuecomment-288362214

If it fails then we might need to make the index of the "token" column unique.

PVince81 replied:
I did a test and apparently the token column has no unique index...

Also I think the unique token problem is not relevant for multiple link shares because the token is always generated by the server and not settable by the user.

However if someone uses the PHP API directly they can do that, but this is again not relevant to this PR here.

Raised https://github.com/owncloud/core/issues/27458 to investigate separately.

PVince81 commented on core/js/sharedialogexpirationview.js:
```diff
@@ -105,30 +84,16 @@
-					$target.addClass('error');
-				}
-			});
+			var expiration = moment($target.val(), 'DD-MM-YYYY').format('YYYY-MM-DD');
```
Not now... this is likely to become a headache -> separate ticket.

We already tried to change this format in the past and it exploded in many places.

PVince81 commented:
@jvillafanez adjustments here https://github.com/owncloud/core/pull/27337/commits/106ff5c23cdfebacf349cac7f26dff78a9207494

PVince81 commented on core/js/sharedialoglinklistview.js:
```diff
@@ -0,0 +1,292 @@
+		_template: undefined,
+
+		events: {
+			'click .addLink': 'onAddButtonClick',
```
Yes, Backbone binds the specified handlers to the class so no need for extra binds.

Commit:
Small tweaks in share link dialog

PVince81 commented:
- [ ] TODO: raise ticket to consolidate and l10nify the date picker date

@felixheidecke adjustments here: 3117e476b689a1dd47bb56dbd5453b1ce45680c0

PVince81 commented:
Uh oh, some test failures:
```
13:43:54 1) Test\Share20\DefaultShareProviderTest::testUpdateLink
13:43:54 Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'INSERT INTO `oc_share` (`share_type`, `uid_initiator`, `uid_owner`, `item_type`, `file_source`, `file_target`, `permissions`, `token`, `parent`) VALUES('3', 'user1', 'user2', 'file', '42', 'target', '31', 'tehtoken', 'some_name')':
13:43:54 
13:43:54 SQLSTATE[HY000]: General error: 1366 Incorrect integer value: 'some_name' for column 'parent' at row 
```

Seems something is shifted... But tests passed locally for me.

I'll have a look.

PVince81 commented:
Unit tests fixed https://github.com/owncloud/core/pull/27337/commits/80ce581ba2f762e149b2570c7a25e4b19eedf42b

PVince81 commented:
- [ ] TODO: set the default name in the dialog to be: folder/file name + counter. Ex: "My Folder 1", "My Folder 2", etc.

PVince81 commented:
Oracle doing wild things:
```
Fatal error:  Call to a member function instanceOfStorage() on null in /var/lib/jenkins/workspace/owncloud-core_core_PR-27337-K2ABFMMCMHW4BOUGXXTF5IXLAJLZUEWLVWTXI7DGCSYR5ZVZBN2Q/lib/private/Files/View.php on line 1899
```

never seen this before...

Commit:
Provide a default deduplicated share link name

Commit:
Set share link name max length to 64 chars

PVince81 commented:
@felixheidecke Instead of just restarting the build I decided to implement the name link suggestion, see 2ef02654f35bb7b3ddcfcfbeeef32d0e9c139b8a

@jvillafanez I know it's not a proper solution but I set the maxlength to 64 in the UI for the name field: d0ef7482923da8726b7b14f276704590fe677527
This should at least prevent accidental overflow through user input.

PVince81 commented:
Okay, so there is something Oracly and weird happening... The same error as before:
```
PHP Fatal error:  Call to a member function instanceOfStorage() on null in /var/lib/jenkins/workspace/owncloud-core_core_PR-27337-K2ABFMMCMHW4BOUGXXTF5IXLAJLZUEWLVWTXI7DGCSYR5ZVZBN2Q/lib/private/Files/View.php on line 1899
```

Not related directly to any of my backend changes, but maybe indirectly causing some weird side-effects...

PVince81 commented:
I've now isolated the backend specific changes that add the "name" column to this separate PR https://github.com/owncloud/core/pull/27448

PVince81 commented:
Still struggling with the new "name" column with Oracle... Expect a longer delay as foretold by the Oracle... https://github.com/owncloud/core/issues/27448

PVince81 commented:
- [ ] BLOCKED by server-side field issue https://github.com/owncloud/core/pull/27448 which itself is blocked by Oracle https://github.com/owncloud/core/pull/27457, reviews needed there first. Then rebase once they are merged.

PVince81 commented:
Merged all blockers, rebased.

👍  for @felixheidecke's code

Let's get this merged, unless you have objections @felixheidecke. Date picker format stuff to be handled separately as it's a different topic.

PVince81 commented:
Cannot unset password when editing, raised here https://github.com/owncloud/core/issues/27481
